tespy/components/customs.py

```diff
@@ -0,0 +1,818 @@
+            k += 1
+        return deriv.tolist()
+
+    def mass_flow_deriv(self):
+        r"""
+        Calculates the partial derivatives for all mass flow balance equations.
+
+        Returns
+        -------
+        deriv : list
+            Matrix with partial derivatives for the mass flow balance
+            equations.
+        """
+        deriv = np.zeros((3, 6 + self.num_vars, self.num_nw_vars))
+        for i in range(self.num_i):
+            deriv[i, i, 0] = 1
+        for j in range(self.num_o):
+            deriv[j, j + i + 1, 0] = -1
+        return deriv.tolist()
+
+    def energy_func(self):
+        r"""
+        Equation for heat exchanger energy balance.
+
+        Returns
+        -------
+        res : float
+            Residual value of equation.
+
+            .. math::
+
+                0 = \dot{m}_{1,in} \cdot \left(h_{1,out} - h_{1,in} \right) +
+                \dot{m}_{2,in} \cdot \left(h_{2,out} - h_{2,in} \right) +
+                \dot{m}_{3,in} \cdot \left(h_{3,out} - h_{3,in} \right)
+        """
+
+        return (self.inl[0].m.val_SI * (self.outl[0].h.val_SI -
+                                        self.inl[0].h.val_SI) +
+                self.inl[1].m.val_SI * (self.outl[1].h.val_SI -
+                                        self.inl[1].h.val_SI) +
+                self.inl[2].m.val_SI * (self.outl[2].h.val_SI -
+                                        self.inl[2].h.val_SI))
+
+    def bus_func(self, bus):
+        r"""
+        Calculates the residual value of the bus function.
+
+        Parameters
+        ----------
+        bus : tespy.connections.bus
+            TESPy bus object.
+
+        Returns
+        -------
+        val : float
+            Residual value of equation.
+
+            .. math::
+
+                val = P \cdot f\left( \frac{P}{P_{ref}}\right)
+
+                P = \dot{m}_{3,in} \cdot \left( h_{3,out} - h_{3,in} \right)
+        """
+        i = self.inl[2].to_flow()
+        o = self.outl[2].to_flow()
+
+        val = i[0] * (o[2] - i[2])
+        if np.isnan(bus.P_ref):
+            expr = 1
+        else:
+            expr = abs(val / bus.P_ref)
+        return val * bus.char.evaluate(expr)
+
+    def bus_deriv(self, bus):
+        r"""
+        Calculates the matrix of partial derivatives of the bus function.
+
+        Parameters
+        ----------
+        bus : tespy.connections.bus
+            TESPy bus object.
+
+        Returns
+        -------
+        mat_deriv : ndarray
+            Matrix of partial derivatives.
+        """
+        deriv = np.zeros((1, 6, self.num_nw_vars))
+        deriv[0, 2, 0] = self.numeric_deriv(self.bus_func, 'm', 2, bus=bus)
+        deriv[0, 2, 2] = self.numeric_deriv(self.bus_func, 'h', 2, bus=bus)
+        deriv[0, 5, 2] = self.numeric_deriv(self.bus_func, 'h', 5, bus=bus)
+        return deriv
+
+    def convergence_check(self, nw):
+        r"""
+        Performs a convergence check.
+        """
```

Comment:
This method is obsolete at the moment. A useful implementation would be changing the enthalpy value of outlet 3 to saturated gas state and the enthalpy value of the geothermal steam outlet to saturated liquid state.

Reply:
Yes...., for that part, we still need to think about it.